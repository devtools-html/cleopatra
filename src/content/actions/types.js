export type ProfileSummaryAction =
  { type: "PROFILE_SUMMARY_PROCESSED", summary: Summary } |
  { type: "PROFILE_SUMMARY_EXPAND", expanded: ExpandedSet } |
  { type: "PROFILE_SUMMARY_COLLAPSE", expanded: ExpandedSet };

export type Action = ProfileSummaryAction;
