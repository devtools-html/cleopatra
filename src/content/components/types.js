// @flow
import type { Component } from 'react';

export type Column = {
  propName: string,
  title: string,
  component?: Class<Component<*, *, *>>,
};

