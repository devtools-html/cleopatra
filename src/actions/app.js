/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { getSelectedTab, getDataSource } from '../selectors/url-state';
import { getTrackThreadHeights } from '../selectors/app';
import { sendAnalytics } from '../utils/analytics';
import type { ThreadIndex } from '../types/profile';
import type { CssPixels } from '../types/units';
import type { Action, ThunkAction } from '../types/store';
import type { TabSlug } from '../app-logic/tabs-handling';
import type { UrlState } from '../types/state';

export function changeSelectedTab(selectedTab: TabSlug): ThunkAction<void> {
  return (dispatch, getState) => {
    const previousTab = getSelectedTab(getState());
    if (previousTab !== selectedTab) {
      sendAnalytics({
        hitType: 'pageview',
        page: selectedTab,
      });
      dispatch({
        type: 'CHANGE_SELECTED_TAB',
        selectedTab,
      });
    }
  };
}

export function profilePublished(hash: string): Action {
  return {
    type: 'PROFILE_PUBLISHED',
    hash,
  };
}

export function changeProfilesToCompare(profiles: string[]): Action {
  return {
    type: 'CHANGE_PROFILES_TO_COMPARE',
    profiles,
  };
}

export function urlSetupDone(): ThunkAction<void> {
  return (dispatch, getState) => {
    dispatch({ type: 'URL_SETUP_DONE' });

    // After the url setup is done, we can successfully query our state about its
    // initial page.
    const dataSource = getDataSource(getState());
    sendAnalytics({
      hitType: 'pageview',
      page: dataSource === 'none' ? 'home' : getSelectedTab(getState()),
    });
    sendAnalytics({
      hitType: 'event',
      eventCategory: 'datasource',
      eventAction: dataSource,
    });
  };
}

export function show404(url: string): Action {
  return { type: 'ROUTE_NOT_FOUND', url };
}

export function changeSidebarOpenState(tab: TabSlug, isOpen: boolean): Action {
  return { type: 'CHANGE_SIDEBAR_OPEN_STATE', tab, isOpen };
}

export function invalidatePanelLayout(): Action {
  return { type: 'INCREMENT_PANEL_LAYOUT_GENERATION' };
}

/**
 * The viewport component provides a hint to use shift to zoom scroll. The first
 * time a user does this, the hint goes away.
 */
export function setHasZoomedViaMousewheel() {
  return { type: 'HAS_ZOOMED_VIA_MOUSEWHEEL' };
}

/**
 * This function is called when a browser navigation event happens. A new UrlState
 * is generated when the window.location is serialized, or the state is pulled out of
 * the history API.
 */
export function updateUrlState(newUrlState: UrlState | null): Action {
  return { type: 'UPDATE_URL_STATE', newUrlState };
}

export const reportTrackThreadHeight = (
  threadIndex: ThreadIndex,
  height: CssPixels
): ThunkAction<void> => (dispatch, getState) => {
  const trackThreadHeights = getTrackThreadHeights(getState());
  const previousHeight = trackThreadHeights[threadIndex];
  if (previousHeight !== height) {
    // Guard against unnecessary dispatches. This could happen frequently.
    dispatch({
      type: 'UPDATE_TRACK_THREAD_HEIGHT',
      height,
      threadIndex,
    });
  }
};

/**
 * This action dismisses the newly published state. This happens when a user first
 * uploads a profile. We only want to remember this when we fist open the profile.
 */
export function dismissNewlyPublished(): Action {
  return { type: 'DISMISS_NEWLY_PUBLISHED' };
}
