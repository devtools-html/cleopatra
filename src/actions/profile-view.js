/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type {
  ProfileSelection, CallTreeFilter, ImplementationFilter,
} from '../types/actions';
import type { Action, ThunkAction } from '../types/store';
import type { ThreadIndex, IndexIntoFuncTable, IndexIntoMarkersTable } from '../types/profile';

/**
 * The actions that pertain to changing the view on the profile, including searching
 * and filtering. Currently the call tree's actions are in this file, but should be
 * split apart. These actions should most likely affect every panel.
 */
export function changeSelectedFuncStack(
  threadIndex: ThreadIndex,
  selectedFuncStack: IndexIntoFuncTable[]
): Action {
  return {
    type: 'CHANGE_SELECTED_FUNC_STACK',
    selectedFuncStack, threadIndex,
  };
}

export function changeSelectedThread(selectedThread: ThreadIndex): Action {
  return {
    type: 'CHANGE_SELECTED_THREAD',
    selectedThread,
  };
}

export function changeThreadOrder(threadOrder: ThreadIndex[]): Action {
  return {
    type: 'CHANGE_THREAD_ORDER',
    threadOrder,
  };
}

export function hideThread(
  threadIndex: ThreadIndex,
  threadOrder: ThreadIndex[],
  hiddenThreads: ThreadIndex[]
): ThunkAction<void> {
  return dispatch => {
    // Do not allow hiding the last thread.
    if (hiddenThreads.length + 1 === threadOrder.length) {
      return;
    }

    dispatch(({
      type: 'HIDE_THREAD',
      threadIndex,
      threadOrder,
      hiddenThreads,
    }: Action));
  };
}

export function showThread(threadIndex: ThreadIndex): Action {
  return {
    type: 'SHOW_THREAD',
    threadIndex,
  };
}

export function changeCallTreeSearchString(searchString: string): Action {
  return {
    type: 'CHANGE_CALL_TREE_SEARCH_STRING',
    searchString,
  };
}

export function changeExpandedFuncStacks(
  threadIndex: ThreadIndex,
  expandedFuncStacks: Array<IndexIntoFuncTable[]>
): Action {
  return {
    type: 'CHANGE_EXPANDED_FUNC_STACKS',
    threadIndex, expandedFuncStacks,
  };
}

export function changeSelectedMarker(
  threadIndex: ThreadIndex,
  selectedMarker: IndexIntoMarkersTable | -1
): Action {
  return {
    type: 'CHANGE_SELECTED_MARKER',
    selectedMarker, threadIndex,
  };
}

export function changeImplementationFilter(implementation: ImplementationFilter): Action {
  return {
    type: 'CHANGE_IMPLEMENTATION_FILTER',
    implementation,
  };
}

export function changeInvertCallstack(invertCallstack: boolean): Action {
  return {
    type: 'CHANGE_INVERT_CALLSTACK',
    invertCallstack,
  };
}

export function changeHidePlatformDetails(hidePlatformDetails: boolean): Action {
  return {
    type: 'CHANGE_HIDE_PLATFORM_DETAILS',
    hidePlatformDetails,
  };
}

export type UpdateProfileSelection = (selection: ProfileSelection) => Action;

export function updateProfileSelection(selection: ProfileSelection): Action {
  return {
    type: 'UPDATE_PROFILE_SELECTION',
    selection,
  };
}

export function addRangeFilter(start: number, end: number): Action {
  return {
    type: 'ADD_RANGE_FILTER',
    start, end,
  };
}

export function addRangeFilterAndUnsetSelection(start: number, end: number): ThunkAction<void> {
  return dispatch => {
    dispatch(addRangeFilter(start, end));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function popRangeFilters(firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_RANGE_FILTERS',
    firstRemovedFilterIndex,
  };
}

export function popRangeFiltersAndUnsetSelection(firstRemovedFilterIndex: number): ThunkAction<void> {
  return dispatch => {
    dispatch(popRangeFilters(firstRemovedFilterIndex));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function addCallTreeFilter(threadIndex: ThreadIndex, filter: CallTreeFilter): Action {
  return {
    type: 'ADD_CALL_TREE_FILTER',
    threadIndex,
    filter,
  };
}

export function popCallTreeFilters(threadIndex: ThreadIndex, firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_CALL_TREE_FILTERS',
    threadIndex,
    firstRemovedFilterIndex,
  };
}
