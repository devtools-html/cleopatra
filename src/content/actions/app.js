// @flow
import type { Action } from './types';

export function changeSelectedTab(selectedTab: string): Action {
  return {
    type: 'CHANGE_SELECTED_TAB',
    selectedTab,
  };
}

export function profilePublished(hash: string): Action {
  return {
    type: 'PROFILE_PUBLISHED',
    hash,
  };
}

export function changeTabOrder(tabOrder: number[]): Action {
  return {
    type: 'CHANGE_TAB_ORDER',
    tabOrder,
  };
}
