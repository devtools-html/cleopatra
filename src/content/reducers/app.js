/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { combineReducers } from 'redux';

import type { Action } from '../actions/types';
import type { State, AppState, AppViewState, Reducer } from './types';

function view(state: AppViewState = { phase: 'INITIALIZING' }, action: Action): AppViewState {
  switch (action.type) {
    case 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_WEB':
      return {
        phase: 'INITIALIZING',
        additionalData: { attempt: action.error.attempt },
      };
    case 'ERROR_RECEIVING_PROFILE_FROM_FILE':
    case 'ERROR_RECEIVING_PROFILE_FROM_ADDON':
    case 'FATAL_ERROR_RECEIVING_PROFILE_FROM_WEB':
      return { phase: 'FATAL_ERROR', error: action.error };
    case 'ROUTE_NOT_FOUND':
      return { phase: 'ROUTE_NOT_FOUND' };
    case 'RECEIVE_PROFILE_FROM_ADDON':
    case 'RECEIVE_PROFILE_FROM_WEB':
    case 'RECEIVE_PROFILE_FROM_FILE':
      return { phase: 'PROFILE' };
    default:
      return state;
  }
}

function isURLSetupDone(state: boolean = false, action: Action) {
  switch (action.type) {
    case '@@urlenhancer/urlSetupDone':
      return true;
    default:
      return state;
  }
}
const appStateReducer: Reducer<AppState> = combineReducers({ view, isURLSetupDone });
export default appStateReducer;

export const getApp = (state: State): AppState => state.app;
export const getView = (state: State): AppViewState => getApp(state).view;
export const getIsURLSetupDone = (state: State): boolean => getApp(state).isURLSetupDone;
