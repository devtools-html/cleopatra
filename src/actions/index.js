/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';
import * as profileUpload from './profile-upload';
import * as shortUrl from './short-url';
import * as urlState from './url-state';

export default Object.assign(
  {},
  app,
  icons,
  profileSummary,
  profileUpload,
  profileView,
  receiveProfile,
  shortUrl,
  timeline,
  urlState
);
