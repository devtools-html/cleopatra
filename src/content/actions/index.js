// @flow
import * as app from './app';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';

export default Object.assign({}, app, profileSummary, profileView, receiveProfile, timeline);
