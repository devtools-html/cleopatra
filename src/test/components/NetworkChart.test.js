/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { render } from 'react-testing-library';
import { Provider } from 'react-redux';

import { changeNetworkSearchString } from '../../actions/profile-view';
import NetworkChart from '../../components/network-chart';
import { changeSelectedTab } from '../../actions/app';
import { ensureExists } from '../../utils/flow';
import {
  TIMELINE_MARGIN_LEFT,
  TIMELINE_MARGIN_RIGHT,
} from '../../app-logic/constants';

import mockCanvasContext from '../fixtures/mocks/canvas-context';
import { storeWithProfile } from '../fixtures/stores';
import {
  getProfileWithMarkers,
  getNetworkMarkers,
} from '../fixtures/profiles/processed-profile';
import { getBoundingBox } from '../fixtures/utils';
import mockRaf from '../fixtures/mocks/request-animation-frame';

import { type NetworkPayload } from '../../types/markers';

const NETWORK_MARKERS = (function() {
  const arrayOfNetworkMarkers = Array(10)
    .fill()
    .map((_, i) => getNetworkMarkers(3 + 0.1 * i, i));
  return [].concat(...arrayOfNetworkMarkers);
})();

function setupWithProfile(profile) {
  const flushRafCalls = mockRaf();
  const ctx = mockCanvasContext();
  jest
    .spyOn(HTMLCanvasElement.prototype, 'getContext')
    .mockImplementation(() => ctx);

  // Ideally we'd want this only on the Canvas and on ChartViewport, but this is
  // a lot easier to mock this everywhere.
  jest
    .spyOn(HTMLElement.prototype, 'getBoundingClientRect')
    .mockImplementation(() =>
      // We're adding the timeline margin to try to get some round numbers in
      // the tests.
      getBoundingBox(200 + TIMELINE_MARGIN_RIGHT + TIMELINE_MARGIN_LEFT, 300)
    );

  const store = storeWithProfile(profile);
  store.dispatch(changeSelectedTab('network-chart'));

  const renderResult = render(
    <Provider store={store}>
      <NetworkChart />
    </Provider>
  );
  const { container } = renderResult;

  function getUrlShorteningParts(): Array<[string, string]> {
    return Array.from(
      container.querySelectorAll('.networkChartRowItemLabel span')
    ).map(node => [node.className, node.textContent]);
  }

  const getPhaseElements = () =>
    Array.from(container.querySelectorAll('.networkChartRowItemBarPhase'));

  const getPhaseElementStyles = () =>
    getPhaseElements().map(element => element.getAttribute('style'));

  function rowItem() {
    return ensureExists(
      container.querySelector('.networkChartRowItem'),
      `Couldn't find the row item in the network chart, with selector .networkChartRowItem`
    );
  }

  return {
    ...renderResult,
    flushRafCalls,
    dispatch: store.dispatch,
    flushDrawLog: () => ctx.__flushDrawLog(),
    getUrlShorteningParts,
    getPhaseElements,
    getPhaseElementStyles,
    rowItem,
  };
}

// create new function to get ProfileWithNetworkMarkers
function setupWithPayload(name: string, payload: NetworkPayload) {
  const profile = getProfileWithMarkers([[name, 0, payload]]);
  const setupResult = setupWithProfile(profile);
  const { flushRafCalls, dispatch } = setupResult;

  dispatch(changeSelectedTab('network-chart'));
  flushRafCalls();

  return setupResult;
}

describe('NetworkChart', function() {
  it('renders NetworkChart correctly', () => {
    const profile = getProfileWithMarkers([...NETWORK_MARKERS]);
    const {
      flushRafCalls,
      dispatch,
      flushDrawLog,
      container,
    } = setupWithProfile(profile);

    dispatch(changeSelectedTab('network-chart'));
    flushRafCalls();

    const drawCalls = flushDrawLog();
    expect(container.firstChild).toMatchSnapshot();
    expect(drawCalls).toMatchSnapshot();
  });
});

describe('NetworkChartRowBar phase calculations', function() {
  it('divides up the different phases of the request with full set of required information', () => {
    const { getPhaseElementStyles } = setupWithPayload(
      'Load 100: https://test.mozilla.org',
      {
        type: 'Network',
        URI: 'https://mozilla.org/img/',
        id: 90001,
        pri: 20,
        count: 10,
        status: 'STATUS_REDIRECT',
        startTime: 10,
        // With an endTime at 99, the profile range goes until 100, which
        // gives integer values for test results.
        endTime: 99,
        domainLookupStart: 20,
        domainLookupEnd: 24,
        connectStart: 25,
        tcpConnectEnd: 26,
        secureConnectionStart: 26,
        connectEnd: 28,
        requestStart: 30,
        responseStart: 60,
        responseEnd: 80,
      }
    );

    expect(getPhaseElementStyles()).toEqual([
      'left: 0px; width: 20px; opacity: 0;',
      'left: 20px; width: 20px; opacity: 0.3333333333333333;',
      'left: 40px; width: 60px; opacity: 0.6666666666666666;',
      'left: 100px; width: 40px; opacity: 1;',
      'left: 140px; width: 38px; opacity: 0;',
    ]);
  });

  it('divides up the different phases of the request with subset of required information', () => {
    const { getPhaseElementStyles } = setupWithPayload(
      'Load 101: https://test.mozilla.org',
      {
        type: 'Network',
        URI: 'https://mozilla.org/img/',
        RedirectURI: 'https://mozilla.org/img/optimized',
        id: 90001,
        pri: 20,
        count: 10,
        status: 'STATUS_REDIRECT',
        startTime: 10,
        endTime: 99,
        requestStart: 20,
        responseStart: 60,
        responseEnd: 80,
      }
    );

    expect(getPhaseElementStyles()).toEqual([
      'left: 0px; width: 20px; opacity: 0;',
      'left: 20px; width: 80px; opacity: 0.6666666666666666;',
      'left: 100px; width: 40px; opacity: 1;',
      'left: 140px; width: 38px; opacity: 0;',
    ]);
  });

  it('takes the full width when there is no details in the payload', () => {
    const { getPhaseElementStyles } = setupWithPayload(
      'Load 101: https://test.mozilla.org',
      {
        type: 'Network',
        URI: 'https://mozilla.org/img/',
        RedirectURI: 'https://mozilla.org/img/optimized',
        id: 90001,
        pri: 20,
        count: 10,
        status: 'STATUS_STOP',
        startTime: 10,
        endTime: 99,
      }
    );
    expect(getPhaseElementStyles()).toEqual([
      'left: 0px; width: 178px; opacity: 1;',
    ]);
  });
});

describe('NetworkChartRowBar URL split', function() {
  function setupForUrl(url: string) {
    return setupWithPayload(`Load 101: ${url}`, {
      type: 'Network',
      URI: url,
      id: 90001,
      pri: 20,
      count: 10,
      status: 'STATUS_STOP',
      startTime: 10,
      endTime: 90,
    });
  }

  it('splits up the url by protocol / domain / path / filename / params / hash', function() {
    const { getUrlShorteningParts } = setupForUrl(
      'https://test.mozilla.org/img/optimized/test.gif?param1=123&param2=321#hashNode2'
    );
    expect(getUrlShorteningParts()).toEqual([
      // Then assert that it's broken up as expected
      ['networkChartRowItemUriOptional', 'https://'],
      ['networkChartRowItemUriRequired', 'test.mozilla.org'],
      ['networkChartRowItemUriOptional', '/img/optimized'],
      ['networkChartRowItemUriRequired', '/test.gif'],
      ['networkChartRowItemUriOptional', '?param1=123&param2=321'],
      ['networkChartRowItemUriOptional', '#hashNode2'],
    ]);
  });

  it('splits properly a url without a path', function() {
    const testUrl = 'https://mozilla.org/';
    const { getUrlShorteningParts } = setupForUrl(testUrl);
    expect(getUrlShorteningParts()).toEqual([
      ['networkChartRowItemUriOptional', 'https://'],
      ['networkChartRowItemUriRequired', 'mozilla.org'],
      ['networkChartRowItemUriRequired', '/'],
    ]);
  });

  it('splits properly a url without a directory', function() {
    const testUrl = 'https://mozilla.org/index.html';
    const { getUrlShorteningParts } = setupForUrl(testUrl);
    expect(getUrlShorteningParts()).toEqual([
      ['networkChartRowItemUriOptional', 'https://'],
      ['networkChartRowItemUriRequired', 'mozilla.org'],
      ['networkChartRowItemUriRequired', '/index.html'],
    ]);
  });

  it('splits properly a url without a filename', function() {
    const testUrl = 'https://mozilla.org/analytics/';
    const { getUrlShorteningParts } = setupForUrl(testUrl);
    expect(getUrlShorteningParts()).toEqual([
      ['networkChartRowItemUriOptional', 'https://'],
      ['networkChartRowItemUriRequired', 'mozilla.org'],
      ['networkChartRowItemUriRequired', '/analytics/'],
    ]);
  });

  it('splits properly a url without a filename and a long directory', function() {
    const testUrl = 'https://mozilla.org/assets/analytics/';
    const { getUrlShorteningParts } = setupForUrl(testUrl);
    expect(getUrlShorteningParts()).toEqual([
      ['networkChartRowItemUriOptional', 'https://'],
      ['networkChartRowItemUriRequired', 'mozilla.org'],
      ['networkChartRowItemUriOptional', '/assets'],
      ['networkChartRowItemUriRequired', '/analytics/'],
    ]);
  });

  it('splits properly a url with a short directory path', function() {
    const testUrl = 'https://mozilla.org/img/image.jpg';
    const { getUrlShorteningParts } = setupForUrl(testUrl);
    expect(getUrlShorteningParts()).toEqual([
      ['networkChartRowItemUriOptional', 'https://'],
      ['networkChartRowItemUriRequired', 'mozilla.org'],
      ['networkChartRowItemUriOptional', '/img'],
      ['networkChartRowItemUriRequired', '/image.jpg'],
    ]);
  });

  it('splits properly a url with a long directory path', function() {
    const testUrl = 'https://mozilla.org/assets/img/image.jpg';
    const { getUrlShorteningParts } = setupForUrl(testUrl);
    expect(getUrlShorteningParts()).toEqual([
      ['networkChartRowItemUriOptional', 'https://'],
      ['networkChartRowItemUriRequired', 'mozilla.org'],
      ['networkChartRowItemUriOptional', '/assets/img'],
      ['networkChartRowItemUriRequired', '/image.jpg'],
    ]);
  });

  it('returns null with an invalid url', function() {
    const { getUrlShorteningParts } = setupForUrl(
      'test.mozilla.org/img/optimized/'
    );
    expect(getUrlShorteningParts()).toEqual([]);
  });
});

describe('NetworkChartRowBar MIME-type filter', function() {
  it('searches for img MIME-Type', function() {
    const { rowItem } = setupWithPayload(
      'Load 101: https://test.mozilla.org/img/optimized/test.png',
      {
        type: 'Network',
        URI: 'https://test.mozilla.org/img/optimized/test.png',
        id: 90001,
        pri: 20,
        count: 10,
        status: 'STATUS_STOP',
        startTime: 10,
        endTime: 90,
      }
    );
    expect(rowItem().classList.contains('networkChartRowItemImg')).toBe(true);
  });

  it('searches for html MIME-Type', function() {
    const { rowItem } = setupWithPayload(
      'Load 101: https://test.mozilla.org/img/optimized/test.html',
      {
        type: 'Network',
        URI: 'https://test.mozilla.org/img/optimized/test.html',
        id: 90001,
        pri: 20,
        count: 10,
        status: 'STATUS_STOP',
        startTime: 10,
        endTime: 90,
      }
    );

    expect(rowItem().classList.contains('networkChartRowItemHtml')).toBe(true);
  });

  it('searches for js MIME-Type', function() {
    const { rowItem } = setupWithPayload(
      'Load 101: https://test.mozilla.org/img/optimized/test.js',
      {
        type: 'Network',
        URI: 'https://test.mozilla.org/img/optimized/test.js',
        id: 90001,
        pri: 20,
        count: 10,
        status: 'STATUS_STOP',
        startTime: 10,
        endTime: 90,
      }
    );

    expect(rowItem().classList.contains('networkChartRowItemJs')).toBe(true);
  });

  it('searches for css MIME-Type', function() {
    const { rowItem } = setupWithPayload(
      'Load 101: https://test.mozilla.org/img/optimized/test.css',
      {
        type: 'Network',
        URI: 'https://test.mozilla.org/img/optimized/test.css',
        id: 90001,
        pri: 20,
        count: 10,
        status: 'STATUS_STOP',
        startTime: 10,
        endTime: 90,
      }
    );

    expect(rowItem().classList.contains('networkChartRowItemCss')).toBe(true);
  });

  it('uses default when no filter applies', function() {
    const { rowItem } = setupWithPayload(
      'Load 101: https://test.mozilla.org/img/optimized/test.xuul',
      {
        type: 'Network',
        URI: 'https://test.mozilla.org/img/optimized/test.xuul',
        id: 90001,
        pri: 20,
        count: 10,
        status: 'STATUS_STOP',
        startTime: 10,
        endTime: 90,
      }
    );

    expect(rowItem().className).toEqual('even networkChartRowItem ');
  });
});

describe('EmptyReasons', () => {
  it("shows a reason when a profile's network markers have been filtered out", () => {
    const profile = getProfileWithMarkers(NETWORK_MARKERS);
    const { dispatch, container } = setupWithProfile(profile);

    dispatch(changeSelectedTab('network-chart'));
    dispatch(changeNetworkSearchString('MATCH_NOTHING'));
    expect(container.querySelector('.EmptyReasons')).toMatchSnapshot();
  });
});
