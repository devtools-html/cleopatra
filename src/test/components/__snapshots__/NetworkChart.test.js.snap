// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmptyReasons shows a reason when a profile's network markers have been filtered out 1`] = `
<div
  className="EmptyReasons"
>
  <h3>
    The 
    network chart
     is empty for “
    Empty
    ”
  </h3>
  <p>
    All network requests were filtered out by the current selection or search term. Try changing the search term, or zooming out.
  </p>
</div>
`;

exports[`NetworkChart renders NetworkChart correctly 1`] = `
<div
  aria-labelledby="network-chart-tab-button"
  className="networkChart"
  id="network-chart-tab"
  role="tabpanel"
>
  <div
    className="markerSettings"
  >
    <div
      className="markerSettingsSpacer"
    />
    <div
      className="markerSettingsSearchbar"
    >
      <label
        className="markerSettingsSearchbarLabel"
      >
        Filter Markers: 
        <form
          className="idleSearchField markerSettingsSearchField"
          onSubmit={[Function]}
        >
          <input
            className="idleSearchFieldInput"
            name="search"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required="required"
            title="Only display markers that match a certain name"
            type="search"
            value=""
          />
          <input
            className="idleSearchFieldButton"
            onClick={[Function]}
            tabIndex={-1}
            type="reset"
          />
        </form>
      </label>
    </div>
  </div>
  <div
    className="treeViewBody"
    onKeyDown={[Function]}
    tabIndex={0}
  >
    <div
      className="treeViewBodyInnerWrapper"
    >
      <div
        className="treeViewBodyInner treeViewBodyInner0"
        style={
          Object {
            "height": "16px",
            "width": undefined,
          }
        }
      >
        <div
          className="treeViewBodyInner treeViewBodyInner0TopSpacer"
          key="-1"
          style={
            Object {
              "height": "0px",
            }
          }
        />
        <div
          className="treeViewBodyInner treeViewBodyInner0InnerChunk"
        >
          <section
            className="even networkChartRowItem "
          >
            <div
              className="networkChartRowItemLabel"
            >
              <span>
                <span
                  className="networkChartRowItemUriOptional"
                >
                  https://
                </span>
                <span
                  className="networkChartRowItemUriRequired"
                >
                  mozilla.org
                </span>
                <span
                  className="networkChartRowItemUriRequired"
                >
                  /
                </span>
              </span>
            </div>
            <div
              className="networkChartRowItemBar"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              style={
                Object {
                  "left": 0,
                  "width": 2.5,
                }
              }
            >
              Array [
                <span
                  className="networkChartRowItemBarInner networkChartRowItemBarRequestQueue"
                  style={
                    Object {
                      "width": "0%",
                    }
                  }
                >
                   
                </span>,
                <span
                  className="networkChartRowItemBarInner networkChartRowItemBarRequest"
                  style={
                    Object {
                      "width": "0%",
                    }
                  }
                >
                   
                </span>,
                <span
                  className="networkChartRowItemBarInner networkChartRowItemBarResponse"
                  style={
                    Object {
                      "width": "100%",
                    }
                  }
                >
                   
                </span>,
              ]
            </div>
          </section>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`NetworkChart renders NetworkChart correctly 2`] = `Array []`;
